### PR TITLE
[mmdet] mmdetection framework

### PR DESCRIPTION
**Pull request by yzbx**
## todo 
- [x] template config file
- [x] docker file
- [x] training
- [x] mining 
- [x] infer
- [x] readme
- [x] multiple task 
- [x] support ymir1.0.0+ 
## code review readme

- mmdet/utils/util_ymir.py

从yolov5中utils/ymir_yolov5.py中复制以下内容
```
class YmirStage(IntEnum)
def get_ymir_process(stage: YmirStage, p: float = 0.0) -> float:
def get_merged_config() -> edict:
```

修改了 
```
def get_weight_file(cfg: edict) -> str:
def write_ymir_training_result(last: bool = False, key_score=None):
```
其它不同之处为删除或新增

- ymir_mining.py

全从yolov5/mining/data_augment.py与yolov5/mining/mining_cald.py中复制过来。

- start.py

从sample_executor中复制过来，并加以修改。

### CODE REVIEW

File: det-mmdetection-tmi/README_ymir.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# det-mmdetection-tmi
+
+`mmdetection` framework for object `det`ection `t`raining/`m`ining/`i`nfer task
+
+# build docker image
+
+```
+docker build -t ymir-executor/mmdet:cuda102-tmi -build-arg SERVER_MODE=dev -f docker/Dockerfile.cuda102 .
```
Comment (fenrir-z) [inline det-mmdetection-tmi/README_ymir.md:8]: SERVER_MODE 除了可以选 dev 以外，还可以选什么？这两个参数的区别是什么？可以在这里写清楚。

File: det-mmdetection-tmi/mmdet/core/evaluation/eval_hooks.py [hunk 2]
```diff
@@ -56,10 +79,15 @@ def _do_evaluate(self, runner):
         results = single_gpu_test(runner.model, self.dataloader, show=False)
         runner.log_buffer.output['eval_iter_num'] = len(self.dataloader)
         key_score = self.evaluate(runner, results)
+        write_ymir_training_result(last=False, key_score=key_score)
         # the key_score may be `None` so it needs to skip the action to save
         # the best checkpoint
         if self.save_best and key_score:
             self._save_ckpt(runner, key_score)
+            # best_score = runner.meta['hook_msgs'].get(
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/core/evaluation/eval_hooks.py:87]: 这个还需要保留吗？

File: det-mmdetection-tmi/mmdet/datasets/ymir.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
+# Copyright (c) OpenMMLab voc.py. All rights reserved.
+# wangjiaxin 2022-04-25
+
+import os.path as osp
+
+# from PIL import Image
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/datasets/ymir.py:6]: 删除本行

File: det-mmdetection-tmi/mmdet/datasets/ymir.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+        with open(ann_file, 'r') as fp:
+            lines = fp.readlines()
+
+        for line in lines:
+            # split any white space
+            img_path, ann_path = line.strip().split()
+            img_path = osp.join(self.data_root, self.img_prefix, img_path)
+            ann_path = osp.join(self.data_root, self.ann_prefix, ann_path)
+            # img = Image.open(img_path)
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/datasets/ymir.py:63]: 删除这两行

File: det-mmdetection-tmi/mmdet/datasets/ymir.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+        # get valid category_id (in annotation, start from 1, arbitary)
+        self.cat_ids = self.coco.get_cat_ids(cat_names=self.CLASSES)
+        # convert category_id to label(train_id, start from 0)
+        self.cat2label = {cat_id: i for i, cat_id in enumerate(self.cat_ids)}
+        self.img_ids = self.coco.get_img_ids()
+        # self.img_ids = list(self.coco.imgs.keys())
+        assert len(self.img_ids) > 0, 'image number must > 0'
+        N = len(self.img_ids)
+        print(f'load {N} image from YMIR dataset')
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/datasets/ymir.py:108]: `print(f'load {len(self.img_ids)} image from YMIR dataset')`
N 就在这里用了一次，可以去掉

File: det-mmdetection-tmi/mmdet/datasets/ymir.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+        data_infos = []
+        total_ann_ids = []
+        for i in self.img_ids:
+            info = self.coco.load_imgs([i])[0]
+            info['filename'] = info['file_name']
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/datasets/ymir.py:114]: 为什么要进行一次这种替换？

File: det-mmdetection-tmi/mmdet/datasets/ymir.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        """Get annotation from TXT file by index.
+
+        Args:
+            idx (int): Index of data.
+
+        Returns:
+            dict: Annotation info of specified index.
+        """
+
+        # img_id = self.data_infos[idx]['id']
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/datasets/ymir.py:140]: 去掉这三行

File: det-mmdetection-tmi/mmdet/datasets/ymir.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        for line in lines:
+            obj = [int(x) for x in line.strip().split(',')[0:5]]
+            # YMIR category id starts from 0, coco from 1
+            category_id, xmin, ymin, xmax, ymax = obj
+            bbox = [xmin, ymin, xmax, ymax]
+            h, w = ymax-ymin, xmax-xmin
+            ignore = 0
+            if self.min_size:
+                assert not self.test_mode
+                w = bbox[2] - bbox[0]
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/datasets/ymir.py:158]: 上面已经计算过一次 w 和 h，这里直接用上面的结果即可

File: det-mmdetection-tmi/mmdet/datasets/ymir.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+        Args:
+            idx (int): Index of data.
+
+        Returns:
+            list[int]: All categories in the image of specified index.
+        """
+
+        cat_ids = []
+        # img_path = self.data_infos[idx]['file_name']
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/datasets/ymir.py:188]: 删除这两行

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    if len(mmdet_cfg.log_config['hooks']) <= 1:
+        mmdet_cfg.log_config['hooks'].append(tensorboard_logger)
+    else:
+        mmdet_cfg.log_config['hooks'][1].update(tensorboard_logger)
+
+    # modify evaluation and interval
+    interval = max(1, mmdet_cfg.runner.max_epochs//30)
+    mmdet_cfg.evaluation.interval = interval
+    mmdet_cfg.evaluation.metric = ymir_cfg.param.get('metric', 'bbox')
+    # Whether to evaluating the AP for each class
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/utils/util_ymir.py:128]: 删除这两行

File: det-mmdetection-tmi/training-template.yaml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+samples_per_gpu: 2
+workers_per_gpu: 2
+max_epochs: 300
+config_file: 'configs/yolox/yolox_nano_8x8_300e_coco.py'
+args_options: ''
+cfg_options: ''
+metric: 'bbox'
+port: 12345
```
Comment (fenrir-z) [inline det-mmdetection-tmi/training-template.yaml:8]: 同上，这个 port 是做什么用的？

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+                raise Exception(
+                    f'unsupported source dataset type {src_dataset_type}')
+
+    # modify model output channel
+    mmdet_model_cfg = mmdet_cfg.model.bbox_head
+    mmdet_model_cfg.num_classes = len(ymir_cfg.param.class_names)
+
+    # modify epochs, checkpoint, tensorboard config
+    if ymir_cfg.param.get('max_epochs', None):
+        mmdet_cfg.runner.max_epochs = ymir_cfg.param.max_epochs
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/utils/util_ymir.py:115]: 如果 ymir_cfg 里面没有设置 max_epochs，mmdet_cfg.runner.max_epochs 的值是否是一个预先已经设置好的默认值？
如果我理解没错的话，从可控的角度出发，应该去掉 if 判断，而直接这样写：

`mmdet_cfg.runner.max_epochs = ymir_cfg.param.get('max_epochs', XXX)`

其中 XXX 是我们自己指定的默认值

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+    mmdet_cfg.evaluation.metric = ymir_cfg.param.get('metric', 'bbox')
+    # Whether to evaluating the AP for each class
+    # mmdet_cfg.evaluation.classwise = True
+    return mmdet_cfg
+
+
+def get_weight_file(cfg: edict) -> str:
+    """
+    return the weight file path by priority
+    find weight file in cfg.param.model_params_path or cfg.param.model_params_path
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/utils/util_ymir.py:136]: `cfg.param.model_params_path` 写了两遍

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        weight_files = [osp.join(work_dir, f) for f in result_files if f.startswith(
+            ('iter_', 'epoch_')) and f.endswith('.pth')]
+
+        if len(weight_files) > 0:
+            newest_weight_file = osp.basename(
+                max(weight_files, key=os.path.getctime))
+
+            stage_name = osp.splitext(newest_weight_file)[0]
+            training_result_file = osp.join(work_dir, 'result.yaml')
+            if osp.exists(training_result_file):
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/utils/util_ymir.py:198]: 有没有必要在 ymir_exec_sdk 里面增加一个 `has_stage_name` 的接口？

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+                with open(training_result_file, 'r') as f:
+                    training_result = yaml.safe_load(f)
+                    model_stages = training_result.get('model_stages', {})
+            else:
+                model_stages = {}
+
+            if stage_name not in model_stages:
+                config_files = [f for f in result_files if f.endswith('.py')]
```
Comment (fenrir-z) [inline det-mmdetection-tmi/mmdet/utils/util_ymir.py:206]: 为什么这些文件也要被保存？

File: det-mmdetection-tmi/start.py [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    monitor.write_monitor_logger(percent=1.0)
+    logging.info("training finished")
+
+
+def _run_mining() -> None:
+    command = 'python3 ymir_mining.py'
+    logging.info(f'start mining: {command}')
+    subprocess.run(command.split(), check=True)
+
+    monitor.write_monitor_logger(percent=1.0)
```
Comment (fenrir-z) [inline det-mmdetection-tmi/start.py:43]: 如果又跑 mining 又跑 infer，当 mining 执行完的时候，进度应该不是 1 才对。

File: det-mmdetection-tmi/tools/train.py [hunk 17]
```diff
@@ -96,8 +97,12 @@ def parse_args():
 
 def main():
```
Comment (fenrir-z) [inline det-mmdetection-tmi/tools/train.py:98]: 这个函数会被调用到吗？

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py (hunk 15 above)
Comment (yzbx) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py:206]: > 为什么这些文件也要被保存？

这些文件是模型配置文件，推理和挖掘都要用到。

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+    # modify dataset config
+    ymir_ann_files = dict(
+        train=ymir_cfg.ymir.input.training_index_file,
+        val=ymir_cfg.ymir.input.val_index_file,
+        test=ymir_cfg.ymir.input.candidate_index_file
+    )
+
+    # validation may augment the image and use more gpu
+    # so set smaller samples_per_gpu for validation
+    samples_per_gpu = ymir_cfg.param.samples_per_gpu
```
[the remark this answers is not in the document]
Comment (yzbx) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py:86]: > 我看到在上面的 template.yaml 中，samples_per_gpu 和 workers_per_gpu 都是被注释掉了的，然后这里又用到了这两个配置项。 如果希望 YMIR 系统感知这个配置项，或者希望 WEB 用户在页面的超参选择里面看到这个配置项，应该在 template.yaml 中把你希望用到的配置项取消注释

这个函数只有training的时候会用到，mining 和 infer不会用到这个函数。

File: det-mmdetection-tmi/infer-template.yaml [hunk 19]
```diff
@@ -0,0 +1,8 @@
+# samples_per_gpu: 2
+# workers_per_gpu: 2
+# max_epochs: 300
+# config_file: 'configs/yolox/yolox_nano_8x8_300e_coco.py'
+# args_options: ''
+cfg_options: ''
+conf_threshold: 0.2
+# port: 12345
```
[the remark this answers is not in the document]
Comment (yzbx) [reply det-mmdetection-tmi/infer-template.yaml]: > 这个配置项是做什么用的？

cfg_options是mmdetection框架支持的一个参数， conf_threshold是推理时的置值度阈值，其它注释的配置已经删除。

File: det-mmdetection-tmi/training-template.yaml (hunk 11 above)
Comment (yzbx) [reply det-mmdetection-tmi/training-template.yaml:10]: > 同上，这个 port 是做什么用的？

port是多gpu训练时用的，可以避免端口冲突的。

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+    epoch_pth_files = [
+        f for f in model_params_path if osp.basename(f).startswith(('epoch_', 'iter_'))]
+    if len(epoch_pth_files) > 0:
+        return max(epoch_pth_files, key=os.path.getctime)
+
+    return ""
+
+
+def write_ymir_training_result(last: bool = False, key_score=None):
```
[the remark this answers is not in the document]
Comment (yzbx) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py]: > 使用 type hint 指定 key_score 的类型，并加上对其作用的描述

已加

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py (hunk 14 above)
Comment (yzbx) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py:215]: > 有没有必要在 ymir_exec_sdk 里面增加一个 `has_stage_name` 的接口？

没有，目前仅mmdetection这个框架用到。

File: det-mmdetection-tmi/start.py (hunk 16 above)
Comment (yzbx) [reply det-mmdetection-tmi/start.py]: > 如果又跑 mining 又跑 infer，当 mining 执行完的时候，进度应该不是 1 才对。 进度 1 的写入，应该拿到 start() 里面去

已修复

File: det-mmdetection-tmi/mmdet/utils/util_ymir.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (fenrir-z) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py:86]: 如果这个函数只在 training 里面用到，那把它放到 ymir_train.py 里面会不会更好？

[the remark this answers is not in the document]
Comment (yzbx) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py:86]: > 如果这个函数只在 training 里面用到，那把它放到 ymir_train.py 里面会不会更好？

会好些

[the remark this answers is not in the document]
Comment (fenrir-z) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py:86]: 我找了一下，这个函数只在 tools/train.py 里面被调用到了，可以考虑放到那里面去，然后在前面加下划线。

[the remark this answers is not in the document]
Comment (yzbx) [reply det-mmdetection-tmi/mmdet/utils/util_ymir.py:86]: > 可以考虑放到那里面去

okay，我也是这么想的，用下划线区分对使用它的人更友好一些。tools/train.py 是mmdetection中的，将Ymir的改动与mmdetection的内容尽量分开。将需要用到的函数放到util_ymir.py 中(目前不到300行)，这样后续维护或者升级mmdetection版本会更容易一些。
